### PR TITLE
lib: lte_link_control: Add support for %XMODEMSLEEP AT notifications

### PR DESCRIPTION
**Pull request by simensrostad**
Add support for %XMODEMSLEEP AT notifications which allows the
application to get updates related to modem sleep.

Closes CIA-101

### CODE REVIEW

File: include/modem/lte_lc.h [hunk 1]
```diff
@@ -189,6 +204,17 @@ struct lte_lc_cell {
...
 
+enum lte_lc_modem_sleep_type {
+	LTE_LC_MODEM_SLEEP_PSM			= 1,
+	LTE_LC_MODEM_SLEEP_RF_INACTIVITY	= 2,	/* For example eDRX */
+	LTE_LC_MODEM_SLEEP_FLIGHT_MODE		= 4,
+};
+
+struct lte_lc_modem_sleep {
+	enum lte_lc_modem_sleep_type type;
+	int64_t time;
```
Comment (jtguggedal) [inline include/modem/lte_lc.h:215]: I think we need a comment here explaining what value infinite sleep is

File: lib/lte_link_control/Kconfig [hunk 2]
```diff
@@ -307,6 +307,28 @@ config LTE_LC_TAU_PRE_WARNING_THRESHOLD_MS
 	help
 		Minimum value of the given T3412 timer that will trigger TAU pre-warnings.
 
+config LTE_LC_MODEM_SLEEP_NOTIFICATIONS
+	bool "Get notifications related to modem sleep"
```
Comment (jtguggedal) [inline lib/lte_link_control/Kconfig:311]: Perhaps simply "Modem sleep notifications"

**Review by jtguggedal — COMMENTED** [review]
Looks good, just a couple of potential doc improvements spotted

**Comment by simensrostad** [comment]
@jtguggedal Review comments addressed. Thanks.

**Review by rlubos — APPROVED** [review]
Looks good,  I couldn't find the `XMODEMSLEEP` in the documentation to verify the command syntax though.

File: lib/lte_link_control/Kconfig [hunk 3]
```diff
@@ -307,6 +307,28 @@ config LTE_LC_TAU_PRE_WARNING_THRESHOLD_MS
 	help
 		Minimum value of the given T3412 timer that will trigger TAU pre-warnings.
 
+config LTE_LC_MODEM_SLEEP_NOTIFICATIONS
+	bool "Modem sleep notifications"
+	help
+		If this option is enabled the application will get notifications when the modem
+		enters and exits sleep. The application will also get pre-warning notifiations
```
Comment (tokangas) [inline lib/lte_link_control/Kconfig:314]:
```suggestion
		enters and exits sleep. The application will also get pre-warning notifications
```

File: lib/lte_link_control/lte_lc_helpers.c [hunk 4]
```diff
@@ -567,3 +567,50 @@ int parse_xt3412(const char *at_response, uint64_t *time)
...
+	}
+
+	err = at_params_unsigned_short_get(&resp_list, AT_XMODEMSLEEP_TYPE_INDEX, &type);
+	if (err) {
+		LOG_ERR("Could not get mode sleep type, error: %d", err);
+		goto clean_exit;
+	}
+	modem_sleep->type = type;
+
+	/* If the <time> parameter is not present sleep time is concidered infininte. */
```
Comment (tokangas) [inline lib/lte_link_control/lte_lc_helpers.c:601]:
```suggestion
	/* If the <time> parameter is not present sleep time is considered infinite. */
```

File: lib/lte_link_control/lte_lc_helpers.h [hunk 5]
```diff
@@ -153,3 +159,14 @@ int parse_cereg(const char *at_response,
  * @return Zero on success or (negative) error code otherwise.
  */
 int parse_xt3412(const char *at_response, uint64_t *time);
+
+/* @brief Parses an XMODEMSLEEP response and extracts the sleep <type> and <time>.
+ *
+ * @note If the <time> parameter -1 after this API call, time shall be concidered infinite.
```
Comment (tokangas) [inline lib/lte_link_control/lte_lc_helpers.h:165]:
```suggestion
 * @note If the <time> parameter -1 after this API call, time shall be considered infinite.
```

**Review by tokangas — APPROVED** [review]
LGTM, just a few typos.

**Review by jtguggedal — APPROVED** [review]
Looks good, thanks!

File: doc/nrf/releases/release-notes-latest.rst [hunk 6]
```diff
@@ -76,6 +76,7 @@ nRF9160
 
     * Added support for %XT3412 AT command notifications, which allows the application to get prewarnings before Tracking Area Updates.
     * Added support for neighbor cell measurements.
+    * Add support for %XMODEMSLEEP AT command notifications which allows the application to get notifications related to modem sleep.
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply doc/nrf/releases/release-notes-latest.rst]: Right, fixed.

**Review by umapraseeda — APPROVED** [review]

**Pull request merged**